src/lib/sqle/spooled-query.service.ts

```diff
@@ -0,0 +1,130 @@
+import { Injectable, Provider, Optional, SkipSelf } from '@angular/core';
+import { HttpErrorResponse } from '@angular/common/http';
+import { TranslateService } from '@ngx-translate/core';
+import { catchError, expand, map, mapTo, skipWhile, switchMap, take, tap } from 'rxjs/operators';
+import { BehaviorSubject, Observable, of, throwError, timer } from 'rxjs';
+import { VantageConnectionService } from './connection.service';
+import { VantageQueryService, IQueryPayload, IQueryResultSet } from './query.service';
+
+interface ISpooledQueryError extends HttpErrorResponse {
+  id: string;
+  detailMessage: string;
+}
+
+const MAX_INTERVAL: number = 10000;
+const BASE_INTERVAL: number = 2000;
+const MAX_RETRIES: number = 2;
+
+enum SpooledQueryState {
+  QUEUED = 'QUEUED',
+  PENDING = 'PENDING',
+  SUBMITTED = 'SUBMITTED',
+  RESPONDING = 'RESPONDING',
+  SPOOLING = 'SPOOLING',
+  RESULT_SET_READY = 'RESULT_SET_READY',
+}
+
+@Injectable()
+export class VantageSpooledQueryService {
+  queryStatus: BehaviorSubject<SpooledQueryState> = new BehaviorSubject<SpooledQueryState>(undefined);
+  queryStatus$: Observable<SpooledQueryState> = this.queryStatus.asObservable();
+  queryStack: string[] = [];
+  constructor(
+    private connectionService: VantageConnectionService,
+    private queryService: VantageQueryService,
+    private translate: TranslateService,
+  ) {}
+
+  cancelLastQuery(): void {
+    const thisQuery: string = this.queryStack.pop();
+
+    return this.deleteSpooledQuery(thisQuery);
+  }
+
+  getRunningInfo(): Observable<SpooledQueryState> {
+    return this.queryStatus$;
+  }
+
+  querySystem(payload: IQueryPayload): Observable<IQueryResultSet> {
+    let currentRuns: number = 1;
+
+    return this.queryService.querySystem(this.connectionService.current, { ...payload, spooledResultSet: true }).pipe(
+      tap((res: any) => this.queryStack.push(res.id)),
```

Comment:
We seem to only pop the queryStack when we cancel a query, not when it finishes.

Reply:
Fixed